src/backend/distributed/planner/multi_logical_planner.c

```diff
@@ -3802,6 +3818,64 @@ SubqueryPushdownMultiNodeTree(Query *queryTree)
 }
 
 
+/*
+ * FlattenJoinVars iterates over provided columnList to identify
+ * Var's that are referenced from join RTE, and reverts back to their
+ * original RTEs.
+ *
+ * This is required because Postgres allows columns to be referenced using
+ * a join alias. Therefore the same column from a table could be referenced
+ * twice using its absolute table name (t1.a), and without table name (a).
+ * This is a problem when one of them is inside the group by clause and the
+ * other is not. Postgres is smart about it to detect that both target columns
+ * resolve to the same thing, and allows a single group by clause to cover
+ * both target entries. We want to make sure we provide correct varno/varattno
+ * values to Postgres so that it could produce valid query.
+ *
+ * Only exception is that, if a join is given an alias name, we do not want to
+ * flatten those var's.
+ *
+ */
+static void
+FlattenJoinVars(List *columnList, Query *queryTree)
+{
+	ListCell *columnCell = NULL;
+	List *rteList = queryTree->rtable;
+
+	foreach(columnCell, columnList)
+	{
+		Var *column = (Var *) lfirst(columnCell);
+		RangeTblEntry *columnRte = NULL;
+
+		Assert(IsA(column, Var));
+
+		/*
+		 * if join has an alias, it is copied over join rte. There is
+		 * no need to find the JoinExpr to check whether it has
+		 * an alias defined.
+		 */
+		columnRte = rt_fetch(column->varno, rteList);
+		if (columnRte->rtekind == RTE_JOIN && columnRte->alias == NULL)
+		{
+			PlannerInfo *root = makeNode(PlannerInfo);
+			Var *normalizedVar = NULL;
+
+			root->parse = (queryTree);
+			root->planner_cxt = CurrentMemoryContext;
+			root->hasJoinRTEs = true;
+
+			normalizedVar = (Var *) flatten_join_alias_vars(root, (Node *) column);
```

Comment:
From Postgresql's source code, which we might consider adding here as well:
```
We use the planner's flatten_join_alias_vars
routine to do the flattening; it wants a PlannerInfo root node, which
fortunately can be mostly dummy.
```

Reply:
ok